quic/s2n-quic-transport/src/stream/controller.rs

```diff
@@ -69,122 +66,181 @@ impl Controller {
     /// If `Poll::Pending` is returned, the waker in the given `context` will be woken
     /// when additional stream capacity becomes available.
-    pub fn poll_local_open_stream(
+    pub fn poll_open_local_stream(
         &mut self,
-        stream_type: StreamType,
+        stream_id: StreamId,
         open_tokens: &mut connection::OpenToken,
         context: &Context,
-    ) -> Poll<()> {
-        match stream_type {
+    ) -> Poll<StreamId> {
+        if cfg!(debug_assertions) {
+            match self.direction(stream_id) {
+                StreamDirection::RemoteInitiatedBidirectional
+                | StreamDirection::RemoteInitiatedUnidirectional => {
+                    panic!("should only be called for locally initiated streams")
+                }
+                _ => (),
+            }
+        }
+
+        let poll = match stream_id.stream_type() {
             StreamType::Bidirectional => self
-                .bidi_controller
-                .local_initiated
+                .local_bidi_controller
                 .poll_open_stream(&mut open_tokens.bidirectional, context),
             StreamType::Unidirectional => self
-                .uni_controller
-                .local_initiated
+                .local_uni_controller
                 .poll_open_stream(&mut open_tokens.unidirectional, context),
+        };
+
+        match poll {
+            Poll::Ready(_) => {
+                // only open streams if there is sufficient capacity based on limits
+                self.on_open_stream(stream_id);
+                Poll::Ready(stream_id)
+            }
+            Poll::Pending => Poll::Pending,
         }
     }
 
     /// This method is called when the remote peer wishes to open a new stream.
     ///
-    /// A `STREAM_LIMIT_ERROR` will be returned if the peer has exceeded the stream limits
-    /// that were communicated by transport parameters or MAX_STREAMS frames.
-    pub fn on_remote_open_stream(&mut self, stream_id: StreamId) -> Result<(), transport::Error> {
-        match stream_id.stream_type() {
+    /// Opening a Stream also opens all lower Streams of the same type. Therefore
+    /// this function validates if there is enough capacity to open all streams.
+    ///
+    /// A `STREAM_LIMIT_ERROR` will be returned if the peer has exceeded the
+    /// stream limits that were communicated by transport parameters or
+    /// MAX_STREAMS frames.
+    pub fn on_open_remote_stream(
+        &mut self,
+        stream_iter: StreamIter,
+    ) -> Result<(), transport::Error> {
+        if cfg!(debug_assertions) {
+            match self.direction(stream_iter.max_stream_id()) {
+                StreamDirection::LocalInitiatedBidirectional
+                | StreamDirection::LocalInitiatedUnidirectional => {
+                    panic!("should only be called for remote initiated streams")
+                }
+                _ => (),
```

Comment:
maybe you could add `is_local/is_remote` or `is_local_initiated/is_remote_initiated` to the `StreamDirection` enum, and then just do something like
```rust
debug_assert!(self.direction(stream_iter.max_stream_id()).is_remote(), "should only be called for remote initiated streams");
```